### PR TITLE
ast: New parser implementation

### LINKED ISSUE #501: Numbers lose precision
Consider:

```
package euler

e = 2.71828182846
```

When querying this `euler.e`, we get `2.718281828`, which is acceptable according to the JSON spec because implementations are free to limit “the range and precision of numbers accepted.”

However, the spec goes on to say: “Since software that implements IEEE 754-2008 binary64 (double precision) numbers [IEEE754] is generally available and widely used, good interoperability can be achieved by implementations that expect no more precision or range than these provide, in the sense that implementations will approximate JSON numbers within the expected precision.”

So for interoperability, I’d expect more precision – possibly exactly IEEE754.

### LINKED ISSUE #1251: Improve parser performance for large-ish JSON values
The parser should be able to efficiently parse JSON values. Currently a representative JSON input can take ~10-20ms to parse. Given that the entire evaluation budget is often on the order of milliseconds, this is too expensive. In many cases this can be avoided by parsing JSON inputs using the Golang `json` package, however, it can be problematic to predict when to use `json` and when to use `ast` for parsing.

# Steps to Reproduce the Problem

Given the following JSON input:

```json
{"uid":"d8fdc6db-44e1-11e9-a10f-021ca99d149a","kind":{"group":"apps","version":"v1beta1","kind":"Deployment"},"resource":{"group":"apps","version":"v1beta1","resource":"deployments"},"namespace":"torin-opa-test","operation":"CREATE","userInfo":{"username":"torin@styra.com","groups":["system:authenticated"]},"object":{"metadata":{"name":"nginx","namespace":"torin-opa-test","uid":"d8fdc047-44e1-11e9-a10f-021ca99d149a","generation":1,"creationTimestamp":"2019-03-12T16:14:01Z","labels":{"run":"nginx"}},"spec":{"replicas":1,"selector":{"matchLabels":{"run":"nginx"}},"template":{"metadata":{"creationTimestamp":null,"labels":{"run":"nginx"}},"spec":{"containers":[{"name":"nginx","image":"nginx","resources":{},"terminationMessagePath":"/dev/termination-log","terminationMessagePolicy":"File","imagePullPolicy":"Always"}],"restartPolicy":"Always","terminationGracePeriodSeconds":30,"dnsPolicy":"ClusterFirst","securityContext":{},"schedulerName":"default-scheduler"}},"strategy":{"type":"RollingUpdate","rollingUpdate":{"maxUnavailable":"25%","maxSurge":"25%"}},"revisionHistoryLimit":2,"progressDeadlineSeconds":600},"status":{}},"oldObject":null}
```

The following Go program takes ~15ms to parse on an Intel Core i7 @ 1.9 GHz.

```golang
package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"time"

	"github.com/open-policy-agent/opa/ast"
)

func main() {

	bs, err := ioutil.ReadFile(os.Args[1])
	if err != nil {
		panic(err)
	}

	input := string(bs)

	for {
		t0 := time.Now()
		_, _, err := ast.ParseStatements("", input)
		dt := time.Since(t0)
		if err != nil {
			panic(err)
		} else {
			fmt.Println("took", dt)
		}
	}
}
```

# Additional Info

This problem has popped up [in](https://github.com/open-policy-agent/opa-istio-plugin/issues/67) [the](https://github.com/mna/pigeon/issues/75) [past](https://itnext.io/optimizing-open-policy-agent-based-kubernetes-authorization-via-go-execution-tracer-7b439bb5dc5b).

### LINKED ISSUE #2198: Leading commas in collections parse but have empty values
When an array/set/object is defined with a leading comma the parser will allow this but the value is always an empty version of the collection.

Ex:
```rego
package example

p := [, 1, 2, 3]
q := {, "a", "b", "c"}
r := {,"a": 1, "b": 2, "c": 3}
```
```
{12:12} /tmp ❯ opa eval -f pretty -d x.rego 'data'
{
  "example": {
    "p": [],
    "q": [],
    "r": {}
  }
} 
```
```
{12:12} /tmp ❯ opa parse x.rego
module
 package
  ref
   data
   "example"
 rule
  head
   p
   array
  body
   expr index=0
    true
 rule
  head
   q
   set
  body
   expr index=0
    true
 rule
  head
   r
   object
  body
   expr index=0
    true
```
^^ The array, set, and object should show the values in the parsed AST

We should either be raising an error or allow parsing them and just ignore the leading comma.

### LINKED ISSUE #2199: Rules can parse with wrapping curly braces
# Expected Behavior
A rule defined like:

```rego
{
    p := 123
}
```

Should not be valid according to the grammar we document

# Actual Behavior

It parses fine, as long as there is only a single set of braces.. ex:

```rego
package example

{
    p := 123
}

q {
    p == 123
}
```
```
{12:18} /tmp ❯ opa eval -d ./x.rego -f pretty 'data'
{
  "example": {
    "p": 123,
    "q": true
  }
}
```

# Additional Info

We should probably raise an error for this case. IMO it is somewhat confusing, and probably not what the author intended as the curly braces mean nothing to the language, they're essentially ignored.

### LINKED ISSUE #2200: In some cases "rules" can parse with dots/hyphens/etc in name
# Expected Behavior
A rule defined like:

```
package example

foo.bar = x {
    x := 123
}
```
Should raise a parse error. This isn't a valid rule name.. Same goes for..

```
package example

foo-bar = x {
    x := 123
}
```
And some others (can toss in `+` and more in there)

# Actual Behavior

The "rule" parses... but isn't what the author probably intended. What you get is:

```
$ opa eval -d /tmp/x.rego -f pretty 'data'
{
  "example": {
    "foo": {
      "bar": 123
    },
    "x": 123
  }
}
```
😞 
# Additional Info

Whats happening is that it is being parsed with a partial rule like `foo[bar] = x { true }` _and_ a rule defined with wrapping braces (see #2199) of `x = 123 { true }`.
```json
{
  "package": {
    "path": [
      {
        "type": "var",
        "value": "data"
      },
      {
        "type": "string",
        "value": "example"
      }
    ]
  },
  "rules": [
    {
      "head": {
        "name": "foo",
        "key": {
          "type": "string",
          "value": "bar"
        },
        "value": {
          "type": "var",
          "value": "x"
        }
      },
      "body": [
        {
          "index": 0,
          "terms": {
            "type": "boolean",
            "value": true
          }
        }
      ]
    },
    {
      "head": {
        "name": "x",
        "value": {
          "type": "number",
          "value": 123
        },
        "assign": true
      },
      "body": [
        {
          "index": 0,
          "terms": {
            "type": "boolean",
            "value": true
          }
        }
      ]
    }
  ]
}

### LINKED ISSUE #2201: Object comprehension can sometimes be parsed as object definition
# Expected Behavior
An expression like:

```rego
package example

p {
    x := {"foo": (1+2) | 1}
}
```
Should be defining an object comprehension (to be inline with behavior of array and set with these ambiguous `|` cases).

# Actual Behavior
It gets parsed as an object where the value for `"foo"` has an `or` call.

```
$ opa parse /tmp/x.rego 'data'
module
 package
  ref
   data
   "example"
 rule
  head
   p
   true
  body
   expr index=0
    ref
     eq
    x
    object
     "foo"
     call
      ref
       or
      call
       ref
        plus
       1
       2
      1
```

# Additional Info

Part of the inconsistency is that if there is no call (the `1+2`) then it _is_ parsed as an object comprehension.

```
package example

p {
    x = {"foo": 1 | 1}
}
```
Yields
```
$ opa parse /tmp/x.rego 'data'
module
 package
  ref
   data
   "example"
 rule
  head
   p
   true
  body
   expr index=0
    ref
     eq
    x
    objectcomprehension
     "foo"
     1
     body
      expr index=0
       1
```

### LINKED ISSUE #2202: Partial rules defining objects must have json string keys in head
# Expected Behavior
In rego a valid object can have _any_ type of key. Ex:

```rego
x := {
  ["a", "b", "c"]: "foo"
}
```
Is perfectly valid.

Building objects with partial rules should therefore allow for constructing the same type of objects like:

```rego
p[["a", "b", "c"]] = "foo" { true }
```

# Actual Behavior
You get a parse error:
```
> p[["a", "b", "c"]] = "foo" { true }
|
1 error occurred: 1:1: rego_parse_error: object key must be string, var, or ref, not array
	p[["a", "b", "c"]] = "foo" { true }
```
But you can define it like...

```
> p[x] = "foo" {
|   x := ["a", "b", "c"]
| }
|
Rule 'p' defined in package repl. Type 'show' to see rules.
```

# Additional Info
These types of objects _are_ somewhat dangerous since you cannot serialize them into JSON for results. However for intermediate data structures in a policy they are supported, and it seems weird that we restrict the syntax.

### LINKED ISSUE #2203: Using the `else` keyword shouldn't always requires a body
# Expected Behavior
Any other rule body can be omitted with an implicit `{ true }` being used for evaluation.

Ex: 
```rego
p = 1
```
_really_ means
```rego
p = 1 {
    true
}
```

So for `else` rule bodies this should also work..

```rego
p = 1 {
    input.x == "foo"
} else = -1
```
Should mean..
```
p = 1 {
    input.x == "foo"
} else = -1 {
    true
}
```

# Actual Behavior

The body for the `else` rule is required:
```
$ opa parse /tmp/x.rego
1 error occurred: /tmp/x.rego:6: rego_parse_error: no match found
	} else = -1
	          ^
```

# Additional Info
We should be able to just apply the same logic we do for any other rule body for these.

### PR DESCRIPTION
**Pull request by tsandall**
This commit replaces the existing PEG generated parser with a parser
implemented by hand. The new parser is more efficient (avoiding old
problems with pathological input cases like {{{{{{{{{}}}}}}}} and
deeply-nested composites in general) and offers better opportunities
for improved error reporting (which has been improved already but
there is still room to grow.)

During the test process of implementing the new parser, we identified
a few issues that were present in the old parser. Those issues are
fixed by this commit.

Fixes #1251
Fixes #501
Fixes #2198
Fixes #2199
Fixes #2200
Fixes #2201
Fixes #2202
Fixes #2203

Co-authored-by: Torin Sandall <torinsandall@gmail.com>
Co-authored-by: Patrick East <east.patrick@gmail.com>

Signed-off-by: Torin Sandall <torinsandall@gmail.com>
Signed-off-by: Patrick East <east.patrick@gmail.com>

### CODE REVIEW

**Comment by tsandall** [comment]
cc @jaspervdj-luminal @srenatus

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,38 @@ project adheres to [Semantic Versioning](http://semver.org/).
...
+#### Benchmark Results
+
+TODO
+
+#### Backwards Compatibility
+
+The new parser contains a small number of backwards incompatible changes that
+correct questionable behaviour from the old parser. These changes affect
+a very small number of actual policies and we feel confident in the decision to
+break backwards compatiblity here.
```
Comment (srenatus) [inline CHANGELOG.md]: typo
```suggestion
break backwards compatibility here.
```

File: ast/internal/scanner/scanner.go [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+
+func isLetter(ch rune) bool {
+	return 'a' <= lower(ch) && lower(ch) <= 'z' || ch == '_'
+}
+
+func isDigit(ch rune) bool {
+	return isDecimal(ch) || ch >= utf8.RuneSelf && unicode.IsDigit(ch)
+}
+
+func isDecimal(ch rune) bool { return '0' <= ch && ch <= '9' }
```
Comment (srenatus) [inline ast/internal/scanner/scanner.go]: This passes gofmt? I wouldn't have thought so 💡

File: ast/rego.peg [hunk 3]
```diff
@@ -1,311 +0,0 @@
-{
```
Comment (srenatus) [inline ast/rego.peg]: 💭 Do we have a valid grammar somewhere? Could be useful to library authors processing Rego code.

File: ast/parser.go [hunk 4]
```diff
@@ -1,5150 +1,1483 @@
...
-			pos:  position{line: 198, col: 1, offset: 5522},
-			expr: &choiceExpr{
-				pos: position{line: 198, col: 11, offset: 5532},
-				alternatives: []interface{}{
-					&ruleRefExpr{
-						pos:  position{line: 198, col: 11, offset: 5532},
-						name: "Number",
-					},
-					&ruleRefExpr{
-						pos:  position{line: 198, col: 20, offs
```
Comment (jaspervdj-luminal) [inline ast/parser.go]: This use of `defer` is incredibly smart.

File: ast/rego.peg (hunk 3 above)
Comment (patrick-east) [reply ast/rego.peg]: There is still one in the language reference documentation https://github.com/open-policy-agent/opa/blob/master/docs/content/policy-reference.md#grammar

File: CHANGELOG.md (hunk 1 above)
Comment (tsandall) [reply CHANGELOG.md]: Fixed.

**Comment by tsandall** [comment]
I've updated the changelog to include the benchstat output against master.

**Pull request merged**